File: binaryTreeCousins.java
// tc o(n) sc o(H)
//simple dfs problem

class Solution {
    TreeNode x_parent, y_parent;
    int x_lvl, y_lvl;
    public boolean isCousins(TreeNode root, int x, int y) {
        if(root == null) {
            return false;
        }
        dfs(root, 0, null, x, y);
        return x_parent != y_parent && x_lvl == y_lvl;
    }
    private void dfs(TreeNode root, int lvl, TreeNode parent, int x, int y) {
        if(root == null) {
            return;
        }
        if(root.val == x) {
            x_lvl = lvl;
            x_parent = parent;
        }
        if(root.val == y) {
            y_lvl = lvl;
            y_parent = parent;
        }
        
        if(x_parent == null || y_parent == null) {
        dfs(root.left, lvl + 1, root, x, y);
        dfs(root.right, lvl + 1, root, x, y);
     }
    }
}

File: BinaryTreeCousins.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    public class BinaryTreeCousins
    {
        /// Time Complexity : O(N) 
        // Space Complexity :O(N) 
        // Did this code successfully run on Leetcode :Yes
        // Any problem you faced while coding this :  No
        public bool IsCousins(TreeNode root, int x, int y)
        {
            Queue<TreeNode> q = new Queue<TreeNode>();
            Queue<TreeNode> qP = new Queue<TreeNode>();
            q.Enqueue(root);
            qP.Enqueue(null);
            while (q.Count != 0)
            {
                int size = q.Count;
                bool x_found = false; TreeNode x_Parent = null;
                bool y_found = false; TreeNode y_Parent = null;
                for (int i = 0; i < size; i++)
                {
                    TreeNode curr = q.Dequeue();
                    TreeNode currP = qP.Dequeue();
                    if (x == curr.val)
                    {
                        x_found = true;
                        x_Parent = currP;
                    }
                    if (y == curr.val)
                    {
                        y_found = true;
                        y_Parent = currP;
                    }
                    if (curr.left != null)
                    {
                        q.Enqueue(curr.left);
                        qP.Enqueue(curr);
                    }

                    if (curr.right != null)
                    {
                        q.Enqueue(curr.right);
                        qP.Enqueue(curr);
                    }


                }
                if (x_found && y_found) return x_Parent != y_Parent;
                if (x_found || y_found) return false;
            }
            return false;

        }
    }
}
